modelator/src/cli/mod.rs

```diff
@@ -74,55 +75,74 @@ pub struct TraceCli {
     /// TLA+ config file with CONSTANTS, INIT and NEXT.
     #[clap(parse(from_os_str), value_hint = ValueHint::FilePath)]
     tla_config: PathBuf,
+    /// The number of traces to generate for a single test.
+    #[clap(long)]
+    num_traces: i32,
 }
 
 impl TraceCli {
     fn run(&self) -> Result<JsonValue, Error> {
-        let runtime = crate::ModelatorRuntime::default();
+        let runtime = {
+            let mut runtime = crate::ModelatorRuntime::default();
+            runtime.model_checker_runtime.traces_per_test = self.num_traces;
+            runtime
+        };
 
         let file_suite =
             TlaFileSuite::from_tla_and_config_paths(&self.tla_module, &self.tla_config)?;
-        let test_names = crate::model::language::Tla::extract_test_names(&file_suite);
 
-        let resp = test_names
+        let test_names = crate::model::language::Tla::extract_test_names(&file_suite)
             .into_iter()
-            .filter(|test_name| allow_test_name(test_name, &self.test))
-            .map(|test_name| {
-                let input_artifacts =
-                    crate::model::language::Tla::generate_test(&test_name, &file_suite)?;
-                let res = {
-                    let mut ret = match self.model_checker {
-                        ModelChecker::Apalache => {
-                            crate::model::checker::Apalache::test(&input_artifacts, &runtime)?
-                        }
-                        ModelChecker::Tlc => {
-                            crate::model::checker::Tlc::test(&input_artifacts, &runtime)?
-                        }
-                    };
-                    ret.0.extends_module_name =
-                        Some(input_artifacts.tla_file.module_name().to_string());
-                    ret
-                };
-                tracing::debug!("Apalache::test output {}", res.0);
-
-                match self.format {
-                    OutputFormat::Json => {
-                        let json_trace =
-                            crate::model::language::Tla::tla_trace_to_json_trace(res.0)?;
-                        tracing::debug!("Tla::tla_trace_to_json_trace output {}", json_trace);
-                        write_json_trace_to_file(
-                            input_artifacts.tla_file.module_name(),
-                            &json_trace,
-                        )
+            .filter(|test_name| allow_test_name(test_name, &self.test));
+
+        let mut res: BTreeMap<String, Result<Vec<JsonValue>, Error>> = BTreeMap::new();
```

Comment:
you convert `Iterator<Item=(K, V)>` to `BTreeMap<K, V>` using `collect()`. So you can move this loop inside iterator combinators as before.

Reply:
btw, I like doing this, because `.collect()` can pass hints about the final allocated memory size. initializing an empty BTreeeMap doesn't help with hinting final allocated memory size.

Reply:
Nice point about the memory hint.

The reason I did this is that I wanted to use `?` inside the iteration, and seems to be a huge pain in rust. People seem to recommend this crate and function

https://docs.rs/itertools/0.10.0/itertools/trait.Itertools.html#method.fold_ok

which looks good. So maybe we can add itertools as a dependency? It seems like a good crate.

Otherwise the rust stdlib offers [try_fold](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.try_fold) which doesn't really solve the problem IMO.

Reply:
Itertools crate also has zip and many other useful things I'm sure.

Reply:
you can use `?` inside combinator, possibly the syntax stabilized in recent stable rust.

just make sure, the closure has a return type `Result`. The following is not allowed.
```
let ys: Vec<_> = v.iter().map(|x| {let y = may_be_err(x)?; y + 1}).collect();
```
Do this,
```
let ys: Vec<Result<_, _>> = v.iter().map(|x| {let y = may_be_err(x)?; Ok(y+1)}).collect();
```
or since Rust iterator support error propagation,
```
let ys: Result<Vec<_>, _> = v.iter().map(|x| {let y = may_be_err(x)?; Ok(y+1)}).collect();
```

Reply:
Ah but the problem here is that we don't want to collect the errors, we just want to escape the whole function if there is any error

Reply:
I see so the error is implicitly propogated. Cool!